File: parallel-programming-with-ms-dotnet/FSharp/Appendix A/RelatedPatterns/Decorator.fs
//===============================================================================
// Microsoft patterns & practices
// Parallel Programming Guide
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// This code released under the terms of the 
// Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
//===============================================================================

namespace Microsoft.Practices.ParallelGuideSamples.RelatedPatterns

open System.Collections.Generic
open System.Drawing
open System.Threading.Tasks

// ------------------------------------------------------------------------------
// Sequential and parallel editor for images
// ------------------------------------------------------------------------------

/// Interface that defines the editor component
type IImageEditor =
    abstract Rotate : RotateFlipType -> seq<Bitmap> -> unit

/// Sequential implementation of the component
type SerialEditor() =
    interface IImageEditor with
        member x.Rotate rotation images =
            for b in images do b.RotateFlip(rotation)

/// Decorated, parallel implementation of the component
type ParallelEditor(decorated : IImageEditor) =
    interface IImageEditor with
        // Modified (decorated) behavior
        member x.Rotate rotation images =
            Parallel.ForEach(images, fun (b:Bitmap) ->
                b.RotateFlip(rotation) ) |> ignore

File: parallel-programming-with-ms-dotnet/CSharp/Appendix A/Examples/Decorator.cs
//===============================================================================
// Microsoft patterns & practices
// Parallel Programming Guide
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// This code released under the terms of the 
// Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
//===============================================================================

using System.Collections.Generic;
using System.Drawing;
using System.Threading.Tasks;

namespace Microsoft.Practices.ParallelGuideSamples.RelatedPatterns
{
    #region Calculate example

    public interface IImageEditor
    {
        void Rotate(RotateFlipType rotation, IEnumerable<Bitmap> images);
    }

    public class SerialEditor : IImageEditor
    {
        public void Rotate(RotateFlipType rotation, IEnumerable<Bitmap> images)
        {
            foreach (Bitmap b in images)
                b.RotateFlip(rotation);
        }
    }

    public class ParallelEditor : IImageEditor
    {
        private IImageEditor decorated;

        public ParallelEditor(IImageEditor decorated)
        {
            this.decorated = decorated;
        }

        // Modified behavior
        public void Rotate(RotateFlipType rotation, IEnumerable<Bitmap> images)
        {
            if (decorated == null)
                return;
            Parallel.ForEach(images, b =>
                {
                    b.RotateFlip(rotation);
                });
        }

        // Additional behavior...
    }

    #endregion
}
